### PR TITLE
[Merged by Bors] - feat: port Data.Countable.Defs

### PR DESCRIPTION
**Pull request by joelriou**
Mathlib SHA: 71ca477041bcd6d7c745fe555dc49735c12944b7

### CODE REVIEW

File: Mathlib/Data/Countable/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Data.Finite.Defs
+import Mathlib.Tactic.MkIffOfInductiveProp
+
+/-!
+# Countable types
+
+In this file we define a typeclass saying that a given `Sort*` is countable. See also `encodable`
```
Comment (semorrison) [inline Mathlib/Data/Countable/Defs.lean:12]:
```suggestion
In this file we define a typeclass saying that a given `Sort _` is countable. See also `Encodable`
```

File: Mathlib/Data/Countable/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  e.symm.injective.countable
+
+theorem Equiv.countable_iff (e : α ≃ β) : Countable α ↔ Countable β :=
+  ⟨fun h => @Countable.of_equiv _ _ h e, fun h => @Countable.of_equiv _ _ h e.symm⟩
+
+instance {β : Type v} [Countable β] : Countable (ULift.{u} β) :=
+  Countable.of_equiv _ Equiv.ulift.symm
+
+/-!
+### Operations on `Sort*`s
```
Comment (semorrison) [inline Mathlib/Data/Countable/Defs.lean:72]:
```suggestion
### Operations on `Sort _`s
```

File: Mathlib/Data/Countable/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  Function.Injective.countable (@Fin.eq_of_veq n)
+
+instance (priority := 100) Finite.to_countable [Finite α] : Countable α :=
+  let ⟨_, ⟨e⟩⟩ := Finite.exists_equiv_fin α
+  Countable.of_equiv _ e.symm
+
+instance : Countable PUnit.{u} :=
+  Subsingleton.to_countable
+
+--@[nolint instance_priority]
```
Comment (semorrison) [inline Mathlib/Data/Countable/Defs.lean:95]: There should be a porting note here.

Comment (joelriou) [reply Mathlib/Data/Countable/Defs.lean]: Thanks for the suggestion. I have added a small note.

File: Mathlib/Data/Countable/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  let ⟨_, ⟨e⟩⟩ := Finite.exists_equiv_fin α
+  Countable.of_equiv _ e.symm
+
+instance : Countable PUnit.{u} :=
+  Subsingleton.to_countable
+
+-- Porting note: removed @[nolint instance_priority] as
+-- there is not (yet?) a linter to complain about the fact
+-- that this instance always applies
+instance PropCat.countable (p : Prop) : Countable p :=
```
Comment (semorrison) [inline Mathlib/Data/Countable/Defs.lean:98]:
```suggestion
instance Prop.countable (p : Prop) : Countable p :=
```
This is just mathport getting confused by `Prop` being capitalised.

File: Mathlib/Data/Countable/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  Subsingleton.to_countable
+
+instance Bool.countable : Countable Bool :=
+  ⟨⟨fun b => cond b 0 1, Bool.injective_iff.2 Nat.one_ne_zero⟩⟩
+
+instance PropCat.countable' : Countable Prop :=
```
Comment (semorrison) [inline Mathlib/Data/Countable/Defs.lean:104]:
```suggestion
instance Prop.countable' : Countable Prop :=
```

File: Mathlib/Data/Countable/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+instance (priority := 100) Finite.to_countable [Finite α] : Countable α :=
+  let ⟨_, ⟨e⟩⟩ := Finite.exists_equiv_fin α
+  Countable.of_equiv _ e.symm
+
+instance : Countable PUnit.{u} :=
+  Subsingleton.to_countable
+
+-- Porting note: removed @[nolint instance_priority] as
+-- there is not (yet?) a linter to complain about the fact
+-- that this instance always applies
```
Comment (semorrison) [inline Mathlib/Data/Countable/Defs.lean:97]: I made https://github.com/leanprover-community/mathlib/pull/17732 to see if there's anything wrong with just satisfying the linter here. Probably it is fine, but I also this doesn't really need to wait.

bors d+

**Comment by joelriou** [comment]
bors merge

**Pull request closed without merging**